pkg/testrig/testrig.go

```diff
@@ -0,0 +1,203 @@
+// for every service, it creates main.go (optionally with DB connection) and a dockerfile to containerize it
+// if the service contains DB tables, it creates sidecar postgres container and creates a schema there
+// finally, it creates docker-compose.yml at the point of all (likely your project's root)
+// that joins all containerized services
+func GenerateRig(fs afero.Fs, templateFile string, outputDir string, modules []*sysl.Module) error {
+	serviceMap, err := readUserData(fs, templateFile)
+	if err != nil {
+		return err
+	}
+
+	applications := make(map[string]*sysl.Application)
+	for _, module := range modules {
+		for appName, app := range module.GetApps() {
+			applications[appName] = app
+		}
+	}
+
+	composeServices := make(map[string]interface{})
+
+	for serviceName, serviceTmpl := range serviceMap {
+		if err := fs.MkdirAll(filepath.Join(outputDir, serviceName), os.ModePerm); err != nil {
+			return err
+		}
+
+		if err := generateMain(fs, serviceName, serviceTmpl, outputDir, appNeedsDb(applications[serviceName])); err != nil {
+			return err
+		}
+
+		if err := generateDockerfile(fs, serviceName, serviceTmpl, outputDir); err != nil {
+			return err
+		}
+
+		if appNeedsDb(applications[serviceName]) {
+			composeServices[serviceName+"_db"] = generateDbService(serviceName)
+			composeServices[serviceName] = generateAppService(
+				&generateAppServiceInput{
+					ServiceName: serviceName,
+					Vars:        serviceTmpl,
+					OutputDir:   outputDir,
+					DependsOn:   []string{serviceName + "_db"},
+				})
+		} else {
+			composeServices[serviceName] = generateAppService(
+				&generateAppServiceInput{
+					ServiceName: serviceName,
+					Vars:        serviceTmpl,
+					OutputDir:   outputDir,
+					DependsOn:   nil,
+				})
+		}
+	}
+
+	return generateCompose(fs, composeServices)
+}
+
+func appNeedsDb(app *sysl.Application) bool {
+	if app == nil {
+		return false
+	}
+	patterns := syslutil.MakeStrSetFromAttr("patterns", app.Attrs)
+	return patterns.Contains("DB")
+}
+
+func readUserData(fs afero.Fs, templateFileName string) (ServiceVarsMap, error) {
+	templateFile, err := fs.Open(templateFileName)
+	if err != nil {
+		return nil, err
+	}
+	defer templateFile.Close()
+
+	byteValue, err := ioutil.ReadAll(templateFile)
+	if err != nil {
+		return nil, err
+	}
+	var vars ServiceVarsMap
+	if err := json.Unmarshal(byteValue, &vars); err != nil {
+		return nil, err
+	}
+	return vars, nil
+}
+
+type generateAppServiceInput struct {
+	ServiceName string
+	Vars        ServiceVars
+	OutputDir   string
+	DependsOn   []string
+}
+
+func generateAppService(input *generateAppServiceInput) map[string]interface{} {
+	return map[string]interface{}{
+		"build": map[string]interface{}{
+			"context":    ".",
+			"dockerfile": filepath.Join(input.OutputDir, input.ServiceName, "Dockerfile"),
+			"args": map[string]interface{}{
+				"GOPROXY": "https://artifactory.gcp.anz/artifactory/go",
```

Comment:
This is an internal URL. It should not be in this public repository. Please remove.

Reply:
oh sorry I keep forgetting this is public one